rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+6. Repeat 5 until the end of epoch.
+7. Repeat 5 - 6 until the stop criteria is reached.
+
+
+### Interfaces
+
+One of our goals is to make `ParameterServerStrategy`’s API consistent with other strategies so that users can easily switch between them. This may be challenging due to the fundamental difference between synchronous and asynchronous training. Therefore, we try to use most of the Distribution Strategy’ APIs but occasionally we will create APIs that still make sense in other strategies as well.
+
+**Note: all proposed APIs in this document are tentative and only valid for our first version. We will revisit them once we get enough feedback from users and the community.**
+
+
+#### Constraints
+
+Function is first-class citizen. Users should only schedule functions instead of running individual ops, in addition to creating variables.
+
+Users can occasionally run individual ops on the client, only for reporting purposes such as printing a metric’s value.
+
+
+#### Schedule/Join Primitives
+
+The `strategy.run` API was initially developed for synchronous training. We propose a new pair of primitives to
+
+
+*   hide the details of load-balancing, fault tolerance and dynamic scheduling
+*   expose the non-blocking semantics to users.
+
+To enable scheduling a function on any worker, we recommend users create the same dataset, but may shuffle differently, on all workers via `strategy.experimental_distribute_datasets_from_function` API.
+
+
+```python
+class ParameterServerStrategyV2:
+  
+  def schedule(self, replica_fn, args=(), kwargs=()):
+    """Schedule the `replica_fn` on all replicas in a sync group (a worker).
+   
+    Schedule the `replica_fn` on all replicas in a sync group that is available,    
+    returns a future of PerReplica immediately if `function` has return values.
+
+    It implements at-least-once semantics for function execution. If a worker
+    fails, it will try to reschedule the function on another replica group or throw     
+    an exception to users. So this method assumes that function execution can be 
+    out of order and function inputs are shared between sync groups. 
+
+    We don't support the cases where `args` or `kwargs` are bound to a specific
+    sync group. We will consider supporting them in the future.
+
+    If there are barriers in `replica_fn`, it is users' responsibility to make
+    sure they won't cause deadlock.
+    """
+    pass
+  
+  def join(self):
+    """Wait until all scheduled functions are finished.
+
+    Raises an error if any of the functions fails to execute. In this case,
+    there is no guarantee that non-failing functions will complete.
+
+    When join() is being called, it is not allowed to call `schedule`.
+    """
+    pass
+
+  def done(self):
+    """Returns True if there is no pending functions to be executed."""
+    pass
+
+  def local_results(self, future_list):
+    """Get concrete values of the future list.
```

Comment:
This is not comprehensible to an outsider. The what of the what?

Reply:
Anything returned from `schedule` can be passed to this method to get the concrete value. The `schedule` just return an opaque object. This is similar to other strategies where users pass the result of `run` to `local_result` to get actual values.

Maybe I don't get your question?